sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -52,23 +53,46 @@ object ConstantFolding extends Rule[LogicalPlan] {
     case _ => false
   }
 
-  def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(AlwaysProcess.fn, ruleId) {
-    case q: LogicalPlan => q.transformExpressionsDownWithPruning(
-      AlwaysProcess.fn, ruleId) {
-      // Skip redundant folding of literals. This rule is technically not necessary. Placing this
-      // here avoids running the next rule for Literal values, which would create a new Literal
-      // object and running eval unnecessarily.
-      case l: Literal => l
-
-      case Size(c: CreateArray, _) if c.children.forall(hasNoSideEffect) =>
-        Literal(c.children.length)
-      case Size(c: CreateMap, _) if c.children.forall(hasNoSideEffect) =>
-        Literal(c.children.length / 2)
-
-      // Fold expressions that are foldable.
-      case e if e.foldable => Literal.create(e.eval(EmptyRow), e.dataType)
+  /**
+   * The method is used to fold the children expression inside a conditional expression which
+   * is not foldable. Some branches may not be evaluated at runtime, so here we should in case of
+   * the exception and leave it to runtime
+   */
+  private def conditionalExpressionFolding(child: Expression): Expression = {
+    if (child.foldable) {
+      try {
+        Literal.create(child.eval(EmptyRow), child.dataType)
```

Comment:
To avoid repeatedly evaluating it, how about we use a TreeNodeTag to remember already-evaluated-but-failed expressions?

Reply:
make sense, add a tag `FAILED_TO_EVALUATED`